lib/Transforms/Scalar/ScalarReplAggregatesHLSL.cpp

```diff
@@ -3531,15 +3554,19 @@ static bool ReplaceMemcpy(Value *V, Value *Src, MemCpyInst *MC,
       } else {
         // Replace Constant with a non-Constant.
         IRBuilder<> Builder(MC);
-        ReplaceConstantWithInst(C, Src, Builder);
+        if (!ReplaceConstantWithInst(C, Src, Builder)) {
+          return false;
+        }
       }
     } else {
       // Try convert special pattern for cbuffer which copy array of float4 to
       // array of float.
       if (!tryToReplaceCBVec4ArrayToScalarArray(V, TyV, Src, TySrc, MC, DL)) {
         IRBuilder<> Builder(MC);
         Src = Builder.CreateBitCast(Src, V->getType());
-        ReplaceConstantWithInst(C, Src, Builder);
+        if (!ReplaceConstantWithInst(C, Src, Builder)) {
```

Comment:
I suspect this is where the additional dominance check is needed as we've created a new Src here. Alternately, we could create the src where it is sure to dominate uses of C by placing it just after Src instead of just after MC.  This would prevent the need to recalculate the dominance tree. Alternately, we could recalculate the tree and check again for total dominance here, but I prefer the first option.

Reply:
When running against the example that causes the crash, we don't actually execute this line. Instead, we enter the call to `ReplaceConstantWithInst` in the first condition block about 10 lines above this one.

Reply:
Yeah. I was wrong about this supposition.